### PR TITLE
[SPARK-3172] [SPARK-3577] improve shuffle spill metrics

### PR DESCRIPTION
**Pull request by sryza**
The posted patch addresses both SPARK-3172 and SPARK-3577.  It renames ShuffleWriteMetrics to WriteMetrics and uses it for tracking all three of shuffle write, spilling on the fetch side, and spilling on the write side (which only occurs during sort-based shuffle).

I'll fix and add tests if people think restructuring the metrics in this way makes sense.

I'm a little unsure about the name shuffleReadSpillMetrics, as spilling happens during aggregation, not read, but I had trouble coming up with something better.

I'm also unsure on what the most useful columns would be to display in the UI - I remember some pushback on adding new columns.  Ultimately these metrics will be most helpful if they can inform users whether and how much they need to increase the number of partitions / increase spark.shuffle.memoryFraction.  Reporting spill time informs users whether spilling is a significant impacting performance.  Reporting memory size can help with understanding how much needs to be done to avoid spilling.

@pwendell any thoughts on this?

### CODE REVIEW

**Comment by kayousterhout** [comment]
@sryza I'm wondering about SPARK-3172 -- is that a common source of confusion / debugging difficulty? Just wondering whether it's worthwhile to fix that, since this patch would be simpler if you could just have a single set of spill metrics for the task.

**Comment by sryza** [comment]
I think SPARK-3172 is probably the more useful portion of this patch.  Ultimately, users want to be able to trace their shuffle resource consumption back to the transformations they used.  A single shuffle spill metric makes it difficult to determine which of the stage boundaries bordering a task (and from there, the transformation that triggered the stage boundary) is the culprit.

File: core/src/main/scala/org/apache/spark/Aggregator.scala [hunk 1]
```diff
@@ -54,14 +55,13 @@ case class Aggregator[K, V, C] (
...
-      val combiners = new ExternalAppendOnlyMap[K, V, C](createCombiner, mergeValue, mergeCombiners)
-      combiners.insertAll(iter)
-      // Update task metrics if context is not null
       // TODO: Make context non optional in a future release
-      Option(context).foreach { c =>
-        c.taskMetrics.memoryBytesSpilled += combiners.memoryBytesSpilled
-        c.taskMetrics.diskBytesSpilled += combiners.diskBytesSpilled
-      }
+      val spillMetrics = Option(context).map(
+        _.taskMetrics.getOrCreateShuffleReadSpillMetrics()).getOrElse(new WriteMetrics())
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/Aggregator.scala]: If we're doing map-side combining, this should be for the shuffle write, right?

**Comment by kayousterhout** [comment]
Ok that makes sense...now my concern is whether it's possible to do that without changing the developer API (see my in-line comment).

A few other things:
-Can you fix the import ordering? :)
-It looks like, in external sorter, right now you just keep track of the spilled bytes.  Can you add the spill time there too?

File: core/src/main/scala/org/apache/spark/Aggregator.scala (hunk 1 above)
Comment (sryza) [inline core/src/main/scala/org/apache/spark/Aggregator.scala]: Ah yeah, looks like this is a copy/paste mistake.  In this path, I think we're always on the map side, so the call should just be `getOrCreateShuffleWriteSpillMetrics`.

**Comment by sryza** [comment]
I rewrote this patch with more of an aim to maintain compatibility than the last version.  This is still not completely done and likely will not past tests.  I'm hoping to get validation on the general approach before polishing.

**Comment by kayousterhout** [comment]
@sryza is the idea here to maintain backward compatibility by re-using the ShuffleWriteMetrics class for spill metrics?

I preferred the old approach, where you defined a DiskWriteMetrics class, and then ShuffleReadMetrics includes "val spillMetrics = new DiskWriteMetrics", and shuffle write metrics includes two DiskWriteMetrics (one for spill and one for normal write).  To make everything backward compatible, I might just write the JsonMetrics in the same way as before for ShuffleWriteMetrics.  I'm just concerned this approach is pretty hard to understand for a new person looking at this.  Happy for others to weigh in though.

**Comment by sryza** [comment]
Both approaches are fine with me.  I've had some discussions recently with @pwendell recently where he's stressed the importance of not breaking developer APIs. Patrick, care to weigh in?

**Comment by nchammas** [comment]
Minor side comment: Can we update the title of this PR to something like:

```
[SPARK-3172] [SPARK-3577] Concise description of changes goes here
```

Right now this PR shows up weird on the [PR dashboard](https://spark-prs.appspot.com/).

**Comment by pwendell** [comment]
This has gone stale so I'd like to close this issue pending further discussion.

**Pull request closed without merging**